### ISSUE TITLE
XMLParser and XMLNode do not load XML correctly

### ISSUE THREAD

**Issue by StasJ** [NCAR/VAPOR#1869]
If you parse the XML file below the resulting `XmlNode` is incorrect. The node `root` only has one child while it is supposed to have two. I believe the problem is that XmlNode assumes that there can only be one child of a given tag whereas that is not the case in XML. For example the old vapor model scene files will not load correctly because of this.

### test.xml
```
<?xml version="1.0" encoding="UTF-8" ?>

<root>
	<time id="0"></time>
	<time id="1"></time>
</root>
```

### Code
```
XmlParser parser;
XmlNode node;
parser.LoadFromFile(&node, "test.xml");
cout << node;
```
### Output
```
<root>
<time id="1" >
</time>
</root>
```

**Comment by clyne** [comment]
That’s because both nodes have the same tag, time. Xmlnode doesn’t
distinguish between nodes with the same name and different attributes.

JC

Sent from my iPhone

On Oct 12, 2019, at 5:33 PM, StasJ <notifications@github.com> wrote:

If you parse the XML file below the resulting XmlNode is incorrect. The
node root only has one child.
test.xml

<?xml version="1.0" encoding="UTF-8" ?>

<root>
	<time id="0"></time>
	<time id="1"></time>
</root>

Code

XmlParser parser;
XmlNode node;
parser.LoadFromFile(&node, "test.xml");
cout << node;

Output

<root>
<time id="1" >
</time>
</root>

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/NCAR/VAPOR/issues/1869?email_source=notifications&email_token=ABTWXCALJ3RWX2L4IDSC3CTQOJNEFA5CNFSM4JAEXLE2YY3PNVWWK3TUL52HS4DFUVEXG43VMWVGG33NNVSW45C7NFSM4HRMYMYQ>,
or unsubscribe

**Comment by StasJ** [comment]
Yes, I figured that out and updated the bug report. Is this not a bug?

**Comment by clyne** [comment]
It’s not a bug.

> On Oct 12, 2019, at 8:38 PM, StasJ <notifications@github.com> wrote:
> 
> Yes, I figured that out and updated the bug report. Is this not a bug?
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> 

John Clyne
National Center for Atmospheric Research
303.497.1236 (w), 303.809.1922 (c)
clyne@ucar.edu

**Comment by StasJ** [comment]
Since the model renderer is supposed to use XML configuration files, we need an XML representation that supports this feature (unless the nodes were to be called "time_1", "time_2", etc. but this is a hack). Should I update `XmlNode` to support this or should I rename it `ParamsNode` and create a new `XMLNode`.

I would suggest renaming the current implementation to `ParamsNode` since it is not a true XMLNode and some of its paradigms don't make sense with it being called an `XMLNode`. It could be constructed from a real `XMLNode` (`new ParamsNode(XMLNode*)`) and have a conversion to an XMLNode for saving it to an XML file.

If we have to keep the current `XMLNode`, I would suggest renaming the current `string GetChild(tag)` and ` int GetNumChildren()` with `string GetUniqueChild` and `int GetNumUniqueChildren()` and adding the functions `vector<string> GetChildren(tag)`, `int GetNumChildrenForTag(tag)`, etc.

**Comment by clyne** [comment]
Changing the definition of a valid graph would probably be opening up a can of worms. I really don’t want to go there. Elsewhere in the application where the number of children is not known a priori we’ve simply used a unique naming convention. E.g. time_0, time_1, etc. 




> On Oct 13, 2019, at 12:58 PM, StasJ <notifications@github.com> wrote:
> 
> Since the model renderer is supposed to use XML configuration files, we need an XML representation that supports this feature (unless the nodes were to be called "time_1", "time_2", etc. but this is a hack). Should I update XmlNode to support this or should I rename it ParamsNode and create a new XMLNode.

> 
> I would suggest renaming the current implementation to ParamsNode since it is not a true XMLNode and some of its paradigms don't make sense with it being called an XMLNode. It could be constructed from a real XMLNode (new ParamsNode(XMLNode*)) and have a conversion to an XMLNode for saving it to an XML file.
> 
> If we have to keep the current XMLNode, I would suggest renaming the current string GetChild(tag) and int GetNumChildren() with string GetUniqueChild and int GetNumUniqueChildren() and adding the functions vector<string> GetChildren(tag), int GetNumChildrenForTag(tag), etc.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> 

John Clyne
National Center for Atmospheric Research
303.497.1236 (w), 303.809.1922 (c)
clyne@ucar.edu

**Comment by StasJ** [comment]
I agree which is why I think we should rename the current implementation of XMLNode to ParamsNode. It is confusing to have a class called XMLNode that does not support a basic feature of XML.
Also, if we keep the current implementation it will be confusing to users that we use a non-standard XML format. This is fine for files that are not really meant to be edited by users like session files but users will be creating the scene files manually.

**Comment by StasJ** [comment]
I agree which is why I think we should rename the current implementation of XMLNode to ParamsNode. It is confusing to have a class called XMLNode that does not support a basic feature of XML.

**Comment by clyne** [comment]
I'd like to close this.
